pageserver/src/tenant/size.rs

```diff
@@ -304,7 +434,8 @@ impl ModelInputs {
                     storage.insert_point(&Some(*timeline_id), "".into(), now, None);
                 }
                 Command::BranchFrom(parent) => {
-                    storage.branch(parent, Some(*timeline_id));
+                    // This branch command may fail if it cannot find a parent to branch from.
+                    storage.branch(parent, Some(*timeline_id))?;
```

Comment:
Did you find a test case where this is possible?

Reply:
Ah yes, I mean some different case apart from the bad sort.

I implemented the alternative fix for the issue (fixing the PartialOrd + Ord impls) here: https://gist.github.com/koivunej/598eb84e03b7e9046ce2136d6bcafe4a

I think the test case should at least be picked up (apologies if you already did).

Reply:
for 2 branches, we can use check parent-child relations directly, but what about this case:
```
upd1- branch a from main at lsn 10
upd2- branch c from b at lsn 10
upd3- branch b from a at lsn 10
```

how would we compare upd1 with upd2 

Or I am overthinking and it is impossible to have this situation with more than 2 BranchFrom updates at the same lsn ?

Reply:
Surely the first sort is no longer needed if using the tree sort version? Though looking at now, it does depend on the order... This makes me think that perhaps just fixing the Ord impl as in my patch is a simpler fix. Perhaps a more general topo-sort could be thought of where `(lsn, branchfrom(timeline_id))` are considered, but in my mind that reduces to what the alternative fix is, and could be just done with a unstable sort.

Well, could be that I am not seeing something ... If it's not needed, perhaps the known-bad PartialOrd, Ord impls on Command and Update should be removed as well.

A note about these test cases, to make then exhaustive: `itertools` would give [`permutations`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.permutations) for us, or `permutohedron` could similarly be used instead of writing out the 4 timeline case (as with 3179 data) does become rather keyboard wrecking and increases the chance of failure.

Reply:
> for 2 branches, we can use check parent-child relations directly, but what about this case:
> (three branches at the same lsn)

You are correct! I think these are allowed by our model of timelines so yeah so the alt-fix approach does not work at all.

Reply:
So, what's the next step then? We keep my current fix?

Reply:
Perhaps the PartialOrd, Ord derives cannot be removed then, I at least can't see a way easily.. But that does not of course mean that there isn't a way. I'd still keep on adding more test cases, even if they are to be deleted later.

Reply:
And yes, keep your fix.